src/OrchardCore/Orchard.Mvc.Core/ModuleProjectRazorFileProvider.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Linq;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.FileProviders.Physical;
+using Microsoft.Extensions.Primitives;
+
+namespace Orchard.Mvc
+{
+    /// <summary>
+    /// This custom <see cref="IFileProvider"/> implementation provides the file contents
+    /// of Module Project Razor files while in a development environment.
+    /// </summary>
+    public class ModuleProjectRazorFileProvider : IFileProvider
+    {
+        private const string MappingFilePath = "obj\\ModuleProjectRazorFiles.map";
+
+        private static ConcurrentDictionary<string, string> _paths;
+        private static object _synLock = new object();
+
+        public ModuleProjectRazorFileProvider(string rootPath)
+        {
+            if (_paths == null)
+            {
+                lock (_synLock)
+                {
+                    if (_paths == null)
+                    {
+                        var path = Path.Combine(rootPath, MappingFilePath);
+
+                        if (File.Exists(path))
+                        {
+                            var paths = File.ReadAllLines(path)
+                                .Select(x => x.Split(new[] { '|' }, StringSplitOptions.RemoveEmptyEntries))
+                                .Where(x => x.Count() == 2).ToDictionary(x => x[1].Replace('\\', '/'), x => x[0]);
+
+                            _paths = new ConcurrentDictionary<string, string>(paths);
+                        }
+                        else
+                        {
+                            _paths = new ConcurrentDictionary<string, string>();
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
if (_paths != null)
{
	return;
}
	
lock (_synLock)
{
	if (_paths != null)
	{
		return;
	}
	
	var path = Path.Combine(rootPath, MappingFilePath);

	if (File.Exists(path))
	{
		_paths = File.ReadAllLines(path)
			.Select(x => x.Split(new[] { '|' }, StringSplitOptions.RemoveEmptyEntries))
			.Where(x => x.Count() == 2)
			.ToDictionary(x => x[1].Replace('\\', '/'), x => x[0]); // IS '\\' '/' XPlatform? Should we be using Path.DirectorySeperator (or whatever it is?)
	}
	else
	{
		_paths = new Dictionary<string, string>();
	}
}